lib/BuildSystem/BuildSystem.cpp

```diff
@@ -2297,6 +2299,212 @@ class ArchiveTool : public Tool {
   }
 };
 
+#pragma mark - StaleFileRemovalTool implementation
+
+class StaleFileRemovalCommand : public Command {
+  std::string description;
+
+  std::vector<std::string> expectedOutputs;
+  std::vector<std::string> filesToDelete;
+  bool computedFilesToDelete = false;
+
+  BuildValue priorValue;
+  bool hasPriorResult = false;
+
+  /// The cached signature, once computed -- 0 is used as a sentinel value.
+  std::atomic<uint64_t> cachedSignature{ 0 };
+
+  virtual uint64_t getSignature() {
+    uint64_t signature = cachedSignature;
+    if (signature != 0)
+      return signature;
+
+    // FIXME: Use a more appropriate hashing infrastructure.
+    using llvm::hash_combine;
+    llvm::hash_code code;
+    for (auto expectedOutput: expectedOutputs) {
+      code = hash_combine(code, expectedOutput);
+    }
+    for (auto fileToDelete: filesToDelete) {
+      code = hash_combine(code, fileToDelete);
+    }
+    code = hash_combine(code, int(computedFilesToDelete));
+    for (auto data : priorValue.toData()) {
+      code = hash_combine(code, data);
+    }
+    code = hash_combine(code, int(hasPriorResult));
```

Comment:
`filesToDelete`, `computedFilesToDelete` and `hasPriorResult` shouldn't be part of the signature.

The signature is used by the build system to detect when a command was "different from last time" and should rerun because it has changed. But the "different" just applies to the things configured in the input file, not to its internal state, if that makes sense.

Its a little moot because we choose to always rerun this command in any case, but we should try and keep it consistent.